### PR TITLE
Support for the differentiable head

### PR DESCRIPTION
**Pull request by blakechi**
This pull request add supports as mentioned in the issue #8.
- added `SetFitHead`, which inherited from Sentence Transformers' [Dense](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Dense.py) to make APIs consistent
- integrated `SetFitHead` to `SetFitModel`
- integrated `SetFitHead` to `SetFitTrainer` (sklearn-based head still works and usage remains the same)
- added new tests to test `SetFitHead` (tested initialization for single/multiple targets, forward, backward)
- added new APIs for `SetFitTrainer`: `trainer.freeze()` and `trainer.unfreeze(keep_body_frozen)`

### CODE REVIEW

File: src/setfit/modeling.py [hunk 1]
```diff
@@ -23,6 +25,43 @@
 MODEL_HEAD_NAME = "model_head.pkl"
 
 
+class SetFitDataset(Dataset):
+    def __init__(self, x, y, tokenizer, max_length=32):
```
Comment (lewtun) [inline src/setfit/modeling.py:29]: Small nit: could we add some types here? For things like `tokenizer`, we can import them with `typing.TYPE_CHECKING` and then wrap the type in quotes " ... "

File: src/setfit/modeling.py [hunk 2]
```diff
@@ -23,6 +25,43 @@
...
+        self.x = x
+        self.y = y
+        self.tokenizer = tokenizer
+        self.max_length = max_length
+
+    def __len__(self):
+        return len(self.x)
+
+    def __getitem__(self, idx):
```
Comment (lewtun) [inline src/setfit/modeling.py:38]: nit:
```suggestion
    def __getitem__(self, idx: int):
```

File: src/setfit/modeling.py [hunk 3]
```diff
@@ -33,6 +72,92 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
             self.model._modules["2"] = models.Normalize()
 
 
+class SetFitHead(models.Dense):
+    """
+    A SetFit head that supports binary/multi-classes logistic regression
```
Comment (lewtun) [inline src/setfit/modeling.py:77]: turbo nit:
```suggestion
    A SetFit head that supports binary and multi-class classification
```

File: src/setfit/modeling.py [hunk 4]
```diff
@@ -33,6 +72,92 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
...
+    for end-to-end training.
+
+    To be compatible with Sentence Transformers, we inherit `Dense` from:
+    https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Dense.py
+
+    Args:
+        in_features (`int`, *optional*):
+            The embedding dimension from the output of the SetFit body. If ignore, will use LazyLinear.
```
Comment (lewtun) [inline src/setfit/modeling.py:85]:
```suggestion
            The embedding dimension from the output of the SetFit body. If`None`, defaults to `LazyLinear`.
```

File: src/setfit/modeling.py [hunk 5]
```diff
@@ -33,6 +72,92 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
...
+
+    def get_config_dict(self):
+        return {
+            "in_features": self.in_features,
+            "out_features": self.out_features,
+            "temperature": self.temperature,
+            "bias": self.bias,
+        }
+
+    def __repr__(self):
```
Comment (lewtun) [inline src/setfit/modeling.py:157]: Nice!

File: src/setfit/modeling.py [hunk 6]
```diff
@@ -41,13 +166,93 @@ def __init__(self, model_body=None, model_head=None, multi_target_strategy=None)
...
+        self,
+        x_train: List[str],
+        y_train: List[int],
+        num_epochs: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        learning_rate: Optional[float] = None,
+        body_learning_rate: Optional[float] = None,
+        l2_weight: Optional[float] = None,
+    ):
+        if isinstance(self.model_head, nn.Module):  # train with pyTorch
```
Comment (lewtun) [inline src/setfit/modeling.py:184]: Very elegant way to preserve backwards compatibility!

File: src/setfit/modeling.py [hunk 7]
```diff
@@ -33,6 +72,92 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
...
+        outputs = None
+        if self.out_features == 1:  # only has one target
+            outputs = torch.sigmoid(logits)
+        else:  # multiple targets
+            outputs = nn.functional.softmax(logits / self.temperature)
+
+        return outputs
+
+    def forward(self, features: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        features.update({"prediction": self._forward(features["sentence_embedding"])})
```
Comment (lewtun) [inline src/setfit/modeling.py:126]: I'm curious about the need for separate `_forward()` and `forward()` methods - is there a reason we can't wrap the logic in a single method?

File: src/setfit/modeling.py [hunk 8]
```diff
@@ -41,13 +166,93 @@ def __init__(self, model_body=None, model_head=None, multi_target_strategy=None)
...
+            criterion = self.model_head.get_loss_fn()
+            optimizer = self._prepare_optimizer(learning_rate, body_learning_rate, l2_weight)
+
+            for epoch_idx in range(num_epochs):
+                for batch in dataloader:
+                    features, labels = batch
+                    optimizer.zero_grad()
+
+                    outputs = self.model_body(features)
+                    predictions = self.model_head._forward(outputs["sentence_embedding"])
```
Comment (lewtun) [inline src/setfit/modeling.py:198]: Depending on the answer to my question about two `forward()` methods, it might also be worth implementing a `__call__` method in `SetFitHead` to simplify this step a bit

File: src/setfit/modeling.py [hunk 9]
```diff
@@ -41,13 +166,93 @@ def __init__(self, model_body=None, model_head=None, multi_target_strategy=None)
...
+        return dataloader
+
+    def _prepare_optimizer(
+        self,
+        learning_rate: float,
+        body_learning_rate: Optional[float],
+        l2_weight: float,
+    ):
+        body_learning_rate = body_learning_rate or learning_rate
+        optimizer = torch.optim.SGD(
```
Comment (lewtun) [inline src/setfit/modeling.py:229]: Would it make sense to use `AdamW` here or did you find `SGD` works better?

File: src/setfit/trainer.py [hunk 10]
```diff
@@ -195,11 +197,57 @@ def call_model_init(self, params: Dict[str, Any] = None):
 
         return model
 
-    def train(self, trial: Union["optuna.Trial", Dict[str, Any]] = None):
+    def freeze(self):
+        """
+        Freeze SetFitModel's differentiable head.
+        Note: call this function only when using the differentiable head.
```
Comment (lewtun) [inline src/setfit/trainer.py:203]: Maybe we could do a small check on the `SetFitModel` and raise a `ValueError` if a user tries to execute this with the wrong model?

File: tests/test_modeling.py [hunk 11]
```diff
@@ -72,3 +109,49 @@ def test_setfit_multilabel_classifier_chain_classifier_model_head():
     )
 
     assert type(model.model_head) is ClassifierChain
+
+
+def test_setfit_single_target_differentiable_head():
```
Comment (lewtun) [inline tests/test_modeling.py:114]: I recently refactored the `test_trainer.py` tests into groups by using `unittest.TestCase`. This has the advantage of allowing us to download groups of models just once vs multiple times per test run. 

Would you be happy to group your tests in a similar way? 

(I realise the existing tests in `test_modeling.py` also need to be grouped - I'll do that later :))

**Review by lewtun — COMMENTED** [review]
Absolutely incredible work on implementing a pure `torch` model @blakechi 🔥 !

I've left a few comments / questions, but overall this is looking really good.

Would you mind sharing a code snippet in the PR description so that others can understand how the new API should work? 

I'm also curious if you tested this implementation with some of the test datasets in our paper? It would be cool to know if (a) this implementation does better than our original one and (b) there are no subtle regressions with the `sklearn` version.

File: src/setfit/modeling.py (hunk 1 above)
Comment (blakechi) [reply src/setfit/modeling.py:29]: okay, I will add them in the next commit

File: src/setfit/modeling.py (hunk 5 above)
Comment (blakechi) [reply src/setfit/modeling.py:157]: Thanks!

File: src/setfit/modeling.py (hunk 7 above)
Comment (blakechi) [reply src/setfit/modeling.py:126]: My original intend is to separate the usages. So `forward()` is for being compatible with Sentence-Transformer, and on the other hand, `_forward()` can be used when users feed embeddings only.

But ya, I think I can merge them together and act correspondly by detecting what format users feed into it. Will fix it in the next commit!

File: src/setfit/modeling.py (hunk 8 above)
Comment (blakechi) [reply src/setfit/modeling.py:198]: Ya, I think by merging `forward()` and `_forward()` together as I answered in your question, I can simplify this step.

File: src/setfit/modeling.py (hunk 9 above)
Comment (blakechi) [reply src/setfit/modeling.py:229]: sorry I used SGD for simplicity. I think maybe we can decide which one to use after running experiments?

File: src/setfit/trainer.py (hunk 10 above)
Comment (blakechi) [reply src/setfit/trainer.py:203]: That's a great idea. Let me add it!

File: tests/test_modeling.py (hunk 11 above)
Comment (blakechi) [reply tests/test_modeling.py:114]: sure no problem 👍

File: src/setfit/trainer.py (hunk 10 above)
Comment (blakechi) [reply src/setfit/trainer.py:203]: Just added it

File: src/setfit/modeling.py (hunk 8 above)
Comment (blakechi) [reply src/setfit/modeling.py:198]: Fixed!

File: src/setfit/modeling.py (hunk 7 above)
Comment (blakechi) [reply src/setfit/modeling.py]: Fixed!

File: src/setfit/modeling.py (hunk 1 above)
Comment (blakechi) [reply src/setfit/modeling.py]: Just added it!

**Comment by blakechi** [comment]
> Absolutely incredible work on implementing a pure `torch` model @blakechi 🔥 !
> 
> I've left a few comments / questions, but overall this is looking really good.
> 
> Would you mind sharing a code snippet in the PR description so that others can understand how the new API should work?
> 
> I'm also curious if you tested this implementation with some of the test datasets in our paper? It would be cool to know if (a) this implementation does better than our original one and (b) there are no subtle regressions with the `sklearn` version.

Thanks! Glad you like it!

Sure, will provide a snippet in the next comment. :)

Sorry I might rush a bit. I wanted to share with you the implementation to make sure the APIs are correct, so I only tested the head by running one step (forward and backward) to check its gradients. 
I will test this implementation on the test datasets. If you have some suggested scripts for me to run, I happy to know!

**Comment by blakechi** [comment]
Here is the snippet for using the differentiable head (partially copied from `README.md`):
```python
from datasets import load_dataset
from sentence_transformers.losses import CosineSimilarityLoss

from setfit import SetFitModel, SetFitTrainer


# Load a dataset from the Hugging Face Hub
dataset = load_dataset("sst2")

# Simulate the few-shot regime by sampling 8 examples per class
num_classes = 2
train_dataset = dataset["train"].shuffle(seed=42).select(range(8 * num_classes))

# Initialize `SetFitModel` with the differentiable head
model = SetFitModel.from_pretrained(
    "sentence-transformers/paraphrase-albert-small-v2",
    use_differentiable_head=True,
    head_params={"out_features": num_classes},
)

# Create trainer
trainer = SetFitTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss_class=CosineSimilarityLoss,
    metric="accuracy",
    batch_size=16,
    num_iterations=20, # The number of text pairs to generate for contrastive learning
    num_epochs=1, # The number of epochs to use for constrastive learning
    column_mapping={"sentence": "text", "label": "label"} # Map dataset columns to text/label expected by trainer
)

# Freeze head
trainer.freeze()

# Do contrastive training
trainer.train(num_epochs=1)

# Unfreeze head
trainer.unfreeze()

# Unfreeze head and freeze body
# trainer.unfreeze(keep_body_frozen=True)

# Train end-to-end
trainer.train(num_epochs=1)
```

File: tests/test_modeling.py (hunk 11 above)
Comment (blakechi) [reply tests/test_modeling.py]: done!

File: src/setfit/modeling.py (hunk 9 above)
Comment (lewtun) [reply src/setfit/modeling.py]: Sounds good to me :)

**Comment by blakechi** [comment]
Hi @lewtun,

Sorry for the late. I was packed by other staffs.

Ya, that sounds good to me. I ran the experiments on the dev set using `pytorch` head and they are similar to the performance on the paper. Below is the results:

| Head | SST-5 | Amazon-CF | CR | Emotion | EnronSpam | AGNews |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| pytorch | 44.4 (2.9) | 39.9 (11.7) | 89.0 (1.0) | 48.4 (5.2) | 89.1 (4.1) | 83.3 (2.9) |

I’m running experiments using `sklearn` head and will report it later once it finishes.

Sure, just resolved it!

File: src/setfit/modeling.py (hunk 9 above)
Comment (blakechi) [reply src/setfit/modeling.py]: I will report the results using SGD later, now the first table (please check below) uses AdamW :)

Comment (blakechi) [reply src/setfit/modeling.py]: Reported the results using `SGD`, `Adam`, and `AdamW`, please see below :)

File: scripts/setfit/cal_score.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import argparse
```
Comment (lewtun) [inline scripts/setfit/cal_score.py:1]: I'm curious what you need this script for? We have a `create_summary_table.py` [script](https://github.com/huggingface/setfit/blob/main/scripts/create_summary_table.py) for aggregating scores across runs, but maybe you needed something else?

File: scripts/setfit/run_fewshot.py [hunk 13]
```diff
@@ -121,7 +129,19 @@ def main():
...
             )
-            trainer.train()
+            if args.classifier == "pytorch":
+                trainer.freeze()
+                trainer.train()
+                trainer.unfreeze(keep_body_frozen=args.keep_body_frozen)
+                trainer.train(
+                    num_epochs=25,
+                    body_learning_rate=1e-5,
+                    learning_rate=args.lr,  # recommand: 1e-2
```
Comment (lewtun) [inline scripts/setfit/run_fewshot.py:139]: turbo nit:
```suggestion
                    learning_rate=args.lr,  # recommend: 1e-2
```

File: src/setfit/modeling.py [hunk 14]
```diff
@@ -13,17 +13,77 @@
...
+
 logging.set_verbosity_info()
 logger = logging.get_logger(__name__)
 
-
+TokenizerOutput = Dict[str, List[int]]
 MODEL_HEAD_NAME = "model_head.pkl"
 
 
+class SetFitDataset(Dataset):
```
Comment (lewtun) [inline src/setfit/modeling.py:35]: I think this would belong better in `setfit.data` - WDYT?

Could we also add a one-line docstring to summarise what it's for?

File: src/setfit/modeling.py [hunk 15]
```diff
@@ -34,52 +94,281 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
...
+                The embeddings from the encoder. If using `dict` format,
+                make sure to store embeddings under the key: 'sentence_embedding'
+                and the outputs will be under the key: 'prediction'.
+            temperature (`float`, *optional*):
+                A logits' scaling factor when using multi-targets (i.e., number of targets more than 1).
+                Will override the temerature given during initialization.
+        Returns:
+        [`Dict[str, torch.Tensor]` or `torch.Tensor`]
+        """
+        is_dict = False  # whether `features` is dict or not
```
Comment (lewtun) [inline src/setfit/modeling.py:162]: nit: I think it helps to be explicit whenever possible 
```suggestion
        is_features_dict = False  # whether `features` is dict or not
```

File: src/setfit/modeling.py [hunk 16]
```diff
@@ -34,52 +94,281 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
...
+        if isinstance(features, dict):
+            assert "sentence_embedding" in features
+            is_dict = True
```
Comment (lewtun) [inline src/setfit/modeling.py:165]:
```suggestion
            is_features_dict = True
```

File: src/setfit/modeling.py [hunk 17]
```diff
@@ -34,52 +94,281 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
...
+
+        x = features["sentence_embedding"] if is_dict else features
+        logits = self.linear(x)
+        if self.out_features == 1:  # only has one target
+            outputs = torch.sigmoid(logits)
+        else:  # multiple targets
+            temperature = temperature or self.temperature
+            outputs = nn.functional.softmax(logits / temperature, dim=-1)
+
+        if is_dict:
```
Comment (lewtun) [inline src/setfit/modeling.py:175]:
```suggestion
        if is_features_dict:
```

**Review by lewtun — COMMENTED** [review]
Thanks so much for sharing the experimental results with the new API - incredible work 🔥 !

This PR is looking really close to being done - I've just left a few small nits and then we can merge 🥳 

I'm excited to see this in action!

File: src/setfit/modeling.py (hunk 14 above)
Comment (blakechi) [reply src/setfit/modeling.py:35]: That's a good idea! Will also add doctoring as well

File: scripts/setfit/cal_score.py (hunk 12 above)
Comment (blakechi) [reply scripts/setfit/cal_score.py:1]: oh haha, I didn't see that script 😂
ok, I will remove `cal_score.py` since it's duplicated to `create_summary_table.py`, thanks!

File: src/setfit/modeling.py (hunk 14 above)
Comment (blakechi) [reply src/setfit/modeling.py]: done!

File: scripts/setfit/cal_score.py (hunk 12 above)
Comment (blakechi) [reply scripts/setfit/cal_score.py]: done!

**Review by lewtun — APPROVED** [review]
Thanks for the final iteration - this looks great so I'm going to merge it now 🔥 !

Amazing contribution and thank you for working on it @blakechi 🤗

**Comment by PhilipMay** [comment]
Good job @blakechi ! Many thanks.

**Comment by blakechi** [comment]
> Thanks for the final iteration - this looks great so I'm going to merge it now 🔥 !
> 
> Amazing contribution and thank you for working on it @blakechi 🤗

I want to thank you for your advice and review as well, those make this PR wonderful! Really like this collaboration! 🤗 @lewtun
